OVNKubernetes CNI IC deployment support

aswinsuryan opened the pull request:
With OVN IC CNI 3 kind of deployment are possible

1)SIngle Zone (Or in non-ic deployment)

This is similar to the current deployment. Here a OVN Kubernetes db service will be running and it will be passed by submariner-operator to the routeagent/network plugin syncer code.

2)MultiZone and more than one node per zone.

In this case there will be no OVN db service will be present. But there will be an endpoint per zone. Submariner operator will pass the comma separated list of IP Port and protocol and the endpoint name to the route agent and networkpluginsyncer. The pod will chose one of them depending on the zone it is running

3)Multizone , one node per zone.

For this deployment there will be no service or endpoints. Submariner-Operator will pass the value local. The pods needs to use socket connection to program ovn.

Fixes: https://github.com/submariner-io/enhancements/issues/191

This PR also passed the NBDB and SBDB values to the submariner-route agent daemon set as route agent will be now responsible for programming the OVN routes and policies.
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

aswinsuryan committed:
OVNKubernetes CNI IC deployment support

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

skitt commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,117 @@ const (
+
+func createClusterNetworkWithEndpoints(endPoints []corev1.Endpoints) *ClusterNetwork {
+	pluginSettings := map[string]string{}
+	var OvnNBDBIPs, OVNSBDBIps string
```
The names should be consistently cased and made more distinguishable, *e.g.*

```suggestion
	var northboundDBIPs, southboundDBIPs string
```

(this is all about OVN so the prefix isn’t all that useful).

aswinsuryan replied:
Done,

skitt commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,117 @@ const (
+		for _, subset := range endPoints[index].Subsets {
+			for _, port := range subset.Ports {
+				if strings.Contains(port.Name, "north") {
+					OvnNBDBIPs += fmt.Sprintf("%s:%s:%s:%s:%d,",
```
The constant part should be included in the format string:

```suggestion
					OvnNBDBIPs += fmt.Sprintf("IC::%s:%s:%s:%d,",
```

This ends up with two colons after “IC”, is that intended?

aswinsuryan replied:
Yes, I thought of splitting it with ":" when we later consume it in submariner.

skitt replied:
But do we need two colons? Wouldn’t `IC:%s:%s:%s:%d,` be sufficient?

aswinsuryan replied:
Oh, that was an oversight, fixed it. Thanks

skitt commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,117 @@ const (
+					OvnNBDBIPs += fmt.Sprintf("%s:%s:%s:%s:%d,",
+						"IC:", endPoints[index].Name, port.Protocol, subset.Addresses[0].IP, OvnNBDBDefaultPort)
+				} else if strings.Contains(port.Name, "south") {
+					OVNSBDBIps += fmt.Sprintf("%s:%s:%s:%s:%d,",
```
Same as above.

skitt commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -82,6 +160,31 @@ func discoverOvnKubernetesNetwork(ctx context.Context, client controllerClient.C
+}
 
-	return clusterNetwork, nil
+func FindEndpoint(ctx context.Context, client controllerClient.Client, endpointNameSpace string) (*corev1.EndpointsList, error) {
```
Does this need to be public?

aswinsuryan replied:
No, changed it to private.

aswinsuryan committed:
Removed SBDB for more than one node per zone deployment

* Removed SBDB configuration as we are not programming SBDB now
will be removed for all the cases once we compltely move to the new
implemnetation

* Fixed issues in naming

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

skitt commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,112 @@ const (
+		for _, subset := range endPoints[index].Subsets {
+			for _, port := range subset.Ports {
+				if strings.Contains(port.Name, "north") {
+					northboundDBIPs += fmt.Sprintf("%s:%s:%s:%s:%d,",
+						"IC", endPoints[index].Name, port.Protocol, subset.Addresses[0].IP, OvnNBDBDefaultPort)
```
```suggestion
					northboundDBIPs += fmt.Sprintf("IC:%s:%s:%s:%d,",
						endPoints[index].Name, port.Protocol, subset.Addresses[0].IP, OvnNBDBDefaultPort)
```

aswinsuryan replied:
Done.

skitt reviewed (approved):
Minor `Printf` tweak.

tpantelis commented on controllers/submariner/route_agent_resources.go:
```diff
@@ -128,4 +134,22 @@ func newRouteAgentDaemonSet(cr *v1alpha1.Submariner, name string) *appsv1.Daemon
 		},
 	}
+
+	if clusterNetwork.PluginSettings != nil {
```
Minor thing but the nil check is unnecessary as Go handles nil map when accessing an element via index.

aswinsuryan replied:
Done.

tpantelis commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,112 @@ const (
+
+	endpointList, err := findEndpoint(ctx, client, ovnPod.Namespace)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error retrieving the endpoints from namespace %q", ovnPod.Namespace)
```
```suggestion
		return nil, errors.Wrapf(err, "error retrieving the endpoints from namespace %q", ovnPod.Namespace)
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,112 @@ const (
+
+	var clusterNetwork *ClusterNetwork
+
+	if endpointList == nil || len(endpointList.Items) == 0 {
+		clusterNetwork, err = createLocalClusterNetwork(), nil
+	} else {
+		clusterNetwork, err = createClusterNetworkWithEndpoints(endpointList.Items), nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
```
`err` is always nil here.

```suggestion
	if endpointList == nil || len(endpointList.Items) == 0 {
		clusterNetwork = createLocalClusterNetwork()
	} else {
		clusterNetwork = createClusterNetworkWithEndpoints(endpointList.Items)
	}
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,112 @@ const (
+	pluginSettings := map[string]string{}
+	var northboundDBIPs string
+
+	for index := 0; index < len(endPoints); index++ {
```
```suggestion
	for index := range endPoints {
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,112 @@ const (
+		}
+	}
+
+	pluginSettings[OvnNBDB] = northboundDBIPs
+
+	return &ClusterNetwork{
+		PluginSettings: pluginSettings,
+	}
```
nit - you can set `PluginSettings` inline.

```suggestion
	return &ClusterNetwork{
		PluginSettings: map[string]string{OvnNBDB, northboundDBIPs},
	}
```

tpantelis replied:
What if `northboundDBIPs` is empty.... Should we still set `OvnNBDB` in the map? Is this an error condition?

aswinsuryan replied:
if endpoint is present there should be a NorthBound and SouthBound IP , else it is an invalid ovn configurartion.

tpantelis commented on pkg/discovery/network/ovnkubernetes.go:
```diff
@@ -39,40 +40,112 @@ const (
+
+func createClusterNetworkWithEndpoints(endPoints []corev1.Endpoints) *ClusterNetwork {
+	pluginSettings := map[string]string{}
+	var northboundDBIPs string
```
I think we should make this an array and use `strings.Join(northboundDBIPs, ",")` afterwards. This avoids repeated concatenation and the trailing comma.

aswinsuryan replied:
done.

aswinsuryan committed:
Refactor the code

Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

tpantelis reviewed (approved):
Need to fix compile errors but otherwise looks good.
